62.SchusslerTest/shader.frag

```diff
@@ -0,0 +1,127 @@
+layout (location = 5) flat in float Alpha;
+layout (location = 6) flat in float AlphaY;
+
+layout (location = 0) out vec4 outColor;
+
+layout (push_constant) uniform PC {
+    layout (offset = 64) vec3 campos;
+    layout (offset = 80) uint testNum;
+} pc;
+
+#define TEST_GGX 1
+#define TEST_BECKMANN 2
+#define TEST_PHONG 3
+#define TEST_AS 4
+#define TEST_OREN_NAYAR 5
+#define TEST_LAMBERT 6
+
+#include <nbl/builtin/glsl/bxdf/brdf/specular/ggx.glsl>
+#include <nbl/builtin/glsl/bxdf/brdf/specular/beckmann.glsl>
+#include <nbl/builtin/glsl/bxdf/brdf/specular/blinn_phong.glsl>
+#include <nbl/builtin/glsl/bxdf/brdf/diffuse/oren_nayar.glsl>
+#include <nbl/builtin/glsl/bxdf/brdf/diffuse/lambert.glsl>
+
+float pdot(vec3 a, vec3 b) {
+    return max(0.0, dot(a,b));
+}
+
+float ap(vec3 wi, vec3 wp, vec3 wg) {
+    return pdot(wi, wp)/pdot(wp, wg);
+}
+
+float at(vec3 wi, vec3 wt, vec3 wp, vec3 wg) {
+    return pdot(wi, wt) * sqrt(1-(pdot(wp,wg)*pdot(wp,wg))) / pdot(wp,wg);
+}
+
+float lambdap(vec3 wi, vec3 wt, vec3 wp, vec3 wg) {
+    return ap(wi, wp, wg) / (ap(wi,wp,wg) + at(wi,wt,wp,wg));
+}
+
+float lambdat(vec3 wi, vec3 wt, vec3 wp, vec3 wg) {
+    return at(wi,wt,wp,wg) / (ap(wi,wp,wg) + at(wi,wt,wp,wg));
+}
+
+float G1(vec3 wo, vec3 wp, float a2) {
+    float dotNV = pdot(wp,wo);
+    float denomC = sqrt(a2 + (1.0f - a2) * dotNV * dotNV) + dotNV;
+
+    return 2.0f * dotNV / denomC;
+}
+
+vec3 evalBRDF(vec3 wi, vec3 wo, vec3 N) {
+    const mat2x3 ior = mat2x3(vec3(1.02,1.3,1.02), vec3(1.0,2.0,1.0));
+    const vec3 albedo = vec3(0.5);
+    float a2 = Alpha;
+    nbl_glsl_IsotropicViewSurfaceInteraction inter_ = nbl_glsl_calcSurfaceInteractionFromViewVector(wi, N );
+    nbl_glsl_AnisotropicViewSurfaceInteraction inter = nbl_glsl_calcAnisotropicInteraction(inter_);
+    nbl_glsl_LightSample _sample = nbl_glsl_createLightSample(wo,inter);
+    nbl_glsl_AnisotropicMicrofacetCache cache = nbl_glsl_calcAnisotropicMicrofacetCache(inter, _sample);
+    vec3 brdf = vec3(0.0);
+    if (pc.testNum == TEST_GGX) {
+        brdf = nbl_glsl_ggx_height_correlated_cos_eval(_sample, inter_, cache.isotropic, ior, a2);
+    } else if (pc.testNum == TEST_BECKMANN) {
+        brdf = nbl_glsl_beckmann_height_correlated_cos_eval(_sample, inter_, cache.isotropic, ior, a2);
+    } else if (pc.testNum == TEST_PHONG) {
+        float n = nbl_glsl_alpha2_to_phong_exp(a2);
+        brdf = nbl_glsl_blinn_phong_cos_eval(_sample, inter_, cache.isotropic, n, ior);
+    } else if (pc.testNum == TEST_AS) {
+        float nx = nbl_glsl_alpha2_to_phong_exp(a2);
+        float aa = 1.0-Alpha;
+        float ny = nbl_glsl_alpha2_to_phong_exp(aa*aa);
+        brdf = nbl_glsl_blinn_phong_cos_eval(_sample, inter, cache, nx, ny, ior);
+    } else if (pc.testNum == TEST_OREN_NAYAR) {
+        brdf = albedo*nbl_glsl_oren_nayar_cos_eval(_sample, inter_, a2);
+    } else if (pc.testNum == TEST_LAMBERT) {
+        brdf = albedo*nbl_glsl_lambertian_cos_eval(_sample);
+    }
```

Comment:
here you should be checking that `inter.isotropic.NdotV` and `_sample.NdotL` are valid (same check like for geometric microfacet you perform for the perturbed one)